src/uast.cc

```diff
@@ -138,83 +220,92 @@ void *UastIteratorNext(UastIterator *iter) {
-  if (!xpathCtx) {
-    goto error2;
-  }
 
-  handler = (xmlGenericErrorFunc)Error;
-  initGenericErrorDefaultFunc(&handler);
+  try {
+    QueryResult queryResult(ctx, node, query, XPATH_NODESET);
 
-  xpathObj = xmlXPathEvalExpression(BAD_CAST(query), xpathCtx);
-  if (!xpathObj) {
-    goto error3;
-  }
+    auto nodeset = queryResult.xpathObj->nodesetval;
+    if (!nodeset) {
+        Error(nullptr, "Unable to get array of result nodes\n");
+        throw std::runtime_error(LastError());
+    }
 
-  // Get array of results
-  result = xpathObj->nodesetval;
-  if (!result) {
-    Error(nullptr, "Unable to get array of results\n");
-    goto error3;
-  }
-  results = result->nodeTab;
-  size = (result) ? result->nodeNr : 0;
-
-  // Sometimes libxml return null results for some invalid queries; do
-  // a first pass to discard them and get the real size for the result
-  for (int i = 0; i < size; i++) {
-    if (results[i] != nullptr && results[i]->_private != nullptr) {
-      ++realSize;
+    try {
+      nodes = new Nodes();
+    } catch(const std::bad_alloc&) {
+      Error(nullptr, "Unable to get memory for nodes\n");
+      throw std::runtime_error(LastError());
     }
-  }
 
-  if (NodesSetSize(nodes, realSize) != 0) {
-    Error(nullptr, "Unable to set nodes size\n");
-    goto error3;
-  }
+    auto results = nodeset->nodeTab;
+    auto size = nodeset->nodeNr;
+    size_t realSize = 0;
+
+    for (int i = 0; i < size; i++) {
+      if (results[i] != nullptr && results[i]->_private != nullptr) {
+        ++realSize;
+      }
+    }
 
-  // Populate array of results
-  for (int i = 0; i < size; i++) {
-    if (results[i] != nullptr && results[i]->_private != nullptr) {
-      nodes->results[nodeIdx++] = results[i]->_private;
+    if (NodesSetSize(nodes, realSize) != 0) {
+      Error(nullptr, "Unable to set nodes size\n");
+      throw std::runtime_error(LastError());
     }
-  }
-  ok = true;
 
-error3:
-  xmlXPathFreeObject(xpathObj);
-error2:
-  xmlXPathFreeContext(xpathCtx);
-error1:
-  xmlFreeDoc(doc);
+    // Populate array of results
+    size_t nodeIdx = 0;
+    for (int i = 0; i < size; i++) {
+      if (results[i] != nullptr && results[i]->_private != nullptr) {
+        nodes->results[nodeIdx++] = results[i]->_private;
+      }
+    }
 
-  if (!ok) {
+    return nodes;
+  } catch (std::runtime_error&) {
     NodesFree(nodes);
-    return nullptr;
   }
-  return nodes;
+
+  return nullptr;
+}
+
+int UastFilterBool(const Uast *ctx, void *node, const char *query) {
+  try {
+    QueryResult queryResult(ctx, node, query, XPATH_BOOLEAN);
+    return queryResult.xpathObj->boolval;
+  } catch (std::runtime_error&) {}
+
+  return -1;
+}
+
+double UastFilterNumber(const Uast *ctx, void *node, const char *query) {
+  try {
+    QueryResult queryResult(ctx, node, query, XPATH_NUMBER);
+    return queryResult.xpathObj->floatval;
+  } catch (std::runtime_error&) {}
+
+  return -1;
```

Comment:
I don't think this is a valid error value.

Reply:
Dammit, it's true.

How about using a flag pointer to indicate success?

```
double UastFilterNumber(const Uast *ctx, void *node, const char *query, bool *ok);
```

Usage:

```
bool ok;
auto mystr = UastFilterNumber(ctx, node, query, &ok);
if (!ok) { ...error...} 
```

Reply:
Yes, that would work. I wonder if it would be better to do the same with the bool case (and use a bool and not an int), for consistency, not sure about string either (in this case a NULL return value is probably best...)
What do you think?

Reply:
Well, this is C, you do what you can, not what you want :)

I like the option of having the bool function return a bool and also have a flag, it's less surprising. I'll do that in the next push. String will be probably expected to return a null on error even by users that don't read the docs.